### DISCUSSION TITLE
Unfinished snippet is making my file get duplicated

### DISCUSSION THREAD

**Discussion by danielo515** [L3MON4D3/LuaSnip#675] [Q&A] [answered]
Hello.
This error is particularly annoying, because it is very hard to reproduce, and, the worst par, very hard to notice, so when I notice I have been adding text to a duplicated area of my code.

The thing is, whenever I leave a snippet "unifinished", this meaning that I do not hit enter until I reach `{0}`, not sure what I do, but then the entire file is copied and pasted at the beginning, something like doing `:%y` and then `ggP`

The snippet that happens me the most (because it is probably the one I use the most) is this:

```lua
  s(
    { trig = "jslog", dscr = "Javascript easier log", regTrig = false },
    fmt(
      [[ //TODO: Remove this debug line
          Js.log2("{}", {}{});
      ]],
      {
        rep(1),
        i(1),
        i(0),
      }
    )
  ),
```

As you can see, it is quite simple, but still very annoying, because I usually don't want to really finish this snippet, but it is mandatory to have one `i(i)` in order to be able to repeat (as far as I know)

Let me know if you need more details to debug this, I don't want to overwhelm with my entire configuration

**Comment by L3MON4D3** [comment, marked as answer]
Ah, yeah that's a problem without solution :/
AFAIK the problem is that the buffer is deleted/it's content replaced completely, and this causes the extmarks which tracks the borders of the i(i) to expand to contain the entire file, and this is then repeated, duplicating the content.
This can be prevented by always jumping through the snippet(most comfortable via `region_check_events`, it jumps to the i(0) once the cursor is outside the region of the snippet), or disabling whatever is replacing the buffer(formatter?)

**Reply by danielo515** [reply]
> the problem is that the buffer is deleted/it's content replaced completely, and this causes the extmarks which tracks the borders of the i(i) to expand to contain the entire file, and this is then repeated, duplicating the content.

Oh, now you mention, it happens (I think) when I save the file, and indeed that runs the formatter, which, if you are right, it works by replacing the entire file.

> This can be prevented by always jumping through the snippet(most comfortable via `region_check_events`, it jumps to the i(0) once the cursor is outside the region of the snippet), or disabling whatever is replacing the buffer(formatter?)

How can I setup such region check events? Because, even without this problem, I find annoying that sometimes, there is an unfinished snippet that, suddenly, makes me jump to a previous position without me expecting that 😅

**Reply by L3MON4D3** [reply]
Ah, in `ls.setup`, check at the bottom of the Readme (delete_check_events is also a good idea)

**Reply by danielo515** [reply]
I was going to setup delete_set_events option just to discovered I already had it configured. It was text changed, so I changed it to be InsertLeave, to see if, when I exit insert mode for saving, I can get rid of this. How is region_check_events better? Gonna try it with InsertLeave too.

-- Some minutes has passed...

Just tried setting both options to InsertLeave, and if I understand the config options correctly, it seems that events that checks the boundaries are being deleted when I leave insert mode? So... My guess is that setting both options to the same value make them predate each other? What would be a sensible config?

**Reply by danielo515** [reply]
Ok, none of the values seems to do what I need, but they will probably improve the situation. It happens to me that, a lot of times I just save the buffer without even moving the cursor, so even if I use InsertLeave events for everything, the cursor is in the same position, and within the boudnaries of the snippet. Is there a way to finish it on insert leave?

**Reply by L3MON4D3** [reply]
Ah, for that you can check #656
Not sure what you mean with "predate each other", they are executed completely independently, `region_check_events` leaves the snippet when the cursor is outside it, and `delete_check_events` removes(/unlinks) the snippet when its text was deleted (which does not happen automatically)
I have `region...` set to `CursorHold,InsertLeave`, and `delete...` to `TextChanged,InsertEnter`, and that works pretty well.

**Reply by danielo515** [reply]
Hey @L3MON4D3 thanks for the linked issue. It works like a charm. Since I added the suggested snippet I didn't experience this problem again.

**Comment by danielo515** [comment]
Oh, I clearly misunderstood the meaning. I thought it was "when to remove
the callbacks that do the region checks", but now it is clear. Do you mean
that you can set them to run during several events? What is the syntax? A
table of events?

**Reply by L3MON4D3** [reply]
Oh just a string, it's inserted verbatim into the autocommand-setup.
For example

```lua
ls.config.setup({
	region_check_events = "CursorHold,InsertLeave",
	delete_check_events = "TextChanged,InsertEnter",
})
```

**Comment by danielo515** [comment]
Perfect 👌 Gonna try that along with the suggested workaround
